### DISCUSSION TITLE
How do I share the token between content script and extension pop up page?

### DISCUSSION THREAD

**Discussion by Mayank-Tripathi32** [open-sauced-craftwork/browser-extension#47] [Q&A] [answered]
Hi, So I am trying to implement the authentication with supabase and have user connect their opensause account using "authenticate" button on extension popup page. But I am running into problems now, I moved the logic to content script but now i am wondering how i can pass the token data to extension so that i can show data there. I have a button like this at end of page which logs the user into using supabase and stores token into localstorage.

![image](https://user-images.githubusercontent.com/70465598/227740543-2428b163-370c-412a-b783-01c2c51f1b34.png)

Currently, The token is saved in localstorage of github page.

![image](https://user-images.githubusercontent.com/70465598/227740393-044c52d3-ba81-43b2-bc5c-c0c1130d2f8a.png)

I would like to have the token available in extension pop page or someway to access it and show its data here:

![image](https://user-images.githubusercontent.com/70465598/227740419-09372402-dac3-430f-bded-5d7526b6fbee.png)

I am using this to authenticate user:

```
signIn: async (res) => {
       const { data, error } = await supabase.auth.signInWithOAuth(res, {
        redirectTo: "https://github.com/",
       });
      if (error) {
        throw error;
      }
      chrome.tabs.create({ url: data.url, active: true }, (tab) => {
        if (tab.status === "complete") {
          setLoad(true);
          chrome.tabs.remove(tab.id);
        }
        sendResponse(request.action + " executed");
      });
    }, 
```

I also was unable to log user directly from popup page as it would just redirect to localstorage after authentication and store token there(localhost:3000) instead of extension local storage. So any advice on how to get that to work is also welcomed.

**Comment by anush008** [comment, marked as answer]
You can save and access the token like so, or at least that's what I think:

1. Execute a content script on the site to retrieve the data using the `localStorage.getItem()` method.

2. Use the `chrome.runtime.sendMessage()` method to send the data back to the extension's background script, since content scripts cannot access `chrome.storage` of the extension directly.

4. Listen for messages from the content script in the background script using the `chrome.runtime.onMessage` event listener.

3. Once you receive the data, store it in the extension's local storage using the `chrome.storage.local.set()` method.

5. You can then access the data from any part of your extension using the `chrome.storage.local.get()` method.

Since you are redirecting the user after authentication, there is a possibility for the content script to run and send the token data to the background script.

**Reply by anush008** [reply]
In my implementation, I've utilized the `chrome.cookies` API to retrieve the `sb-access-token` cookie, which was then used to access session information from the OpenSauced [authentication service](https://api.opensauced.pizza/#/Authentication%20service/checkAuthSession). This approach eliminates the need for SupaBase as a dependency, allowing for greater flexibility and ease of use. It'll enable you to access various protected services of the OpenSauced API like casting votes, adding highlights etc.
